pkg/compiler/compiler.go

```diff
@@ -125,12 +129,12 @@ type compiler struct {
 
 	unsupported  map[string]bool
 	resources    map[string]*ast.Resource
-	typedefs     map[string]*ast.TypeDef
+	typedefs     map[string]map[int]*ast.TypeDef
```

Comment:
Frankly overloading based on number of arguments is not the direction I would like the language to evolve to.
The only current use (fileoff) makes the overloads semantically equivalent, but it's not required by the primitive. If users will actually start to use this for semantically different things, it looks like recipe for confusion. If the things are different, it would be reasonable to give them different names.
We currently have optional arguments and these are limited to just 2 cases: base type and optional mark. But this feature is not supported for templates, and this seems to be the root problem that does not allow us to describe fileoff.
Making the language more consistent and features more orthogonal looks like a better way to go to me. Namely, making optional base types and opt mark work for templates too. This would allow to express more similar cases. E.g.:
```
sys/linux/bpf.txt:# NEED: opt modified on typedefs, this could be btf_type_id[opt]
```

The reasonable syntax is probably:
`type fileoff[BASE] BASE`
And we make `BASE` template arg have special meaning (prohibited in other positions). `BASE` needs to be provided or not based on the same rules as for normal types. Later this may be extended to `opt` as well:
`type fileoff[BASE, OPT] BASE[OPT]`

Or, option 2 (much simpler), just do: fileoff32 and fileoff64, I think that's just 2 cases we ever need :)

Reply:
> The reasonable syntax is probably:
> `type fileoff[BASE] BASE`
> And we make `BASE` template arg have special meaning (prohibited in other positions). `BASE` needs to be provided or not based on the same rules as for normal types. Later this may be extended to `opt` as well:
> `type fileoff[BASE, OPT] BASE[OPT]`

I'm not sure I understand what you're proposing. How would we express the default base type (`int32`) for `fileoff`? Would type templates always default to a `int32` base type??

> Or, option 2 (much simpler), just do: fileoff32 and fileoff64, I think that's just 2 cases we ever need :)

There's also `fileoff[intptr]`: https://github.com/google/syzkaller/blob/c59a7cd871bf29d123481b2e5b0bd739b064f15f/sys/linux/sys.txt#L90

Reply:
> How would we express the default base type (int32) for fileoff?

The default base type will then be intptr (only in syscall args) as it is for any other types.

> There's also fileoff[intptr]:

Well, I don't know, either fileoffptr, or probably just fileoff64.

Reply:
Please keep the bool's for now. If we require writing bool[int8], then it's not shorter than int8[0:1]. Or you have something else in mind?